### PR TITLE
[Merged by Bors] - Rework manual event iterator so we can actually name the type

### PR DESCRIPTION
**Pull request by Aceeri**
# Objective
- Be able to name the type that `ManualEventReader::iter/iter_with_id` returns and `EventReader::iter/iter_with_id` by proxy.
  Currently for the purpose of https://github.com/bevyengine/bevy/pull/5719

## Solution
- Create a custom `Iterator` type.

### CODE REVIEW

**Review by afonsolage — APPROVED** [review]

File: crates/bevy_ecs/src/event.rs [hunk 1]
```diff
@@ -385,57 +360,95 @@ impl<E: Event> ManualEventReader<E> {
...
+    /// Iterate over only the events.
+    pub fn without_id(self) -> ManualEventIterator<'a, E> {
+        ManualEventIterator { iter: self }
+    }
+}
+
+impl<'a, E: Event> Iterator for ManualEventIteratorWithId<'a, E> {
+    type Item = (&'a E, EventId<E>);
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.range.next().and_then(|id| self.events.get_event(id)) {
```
Comment (cart) [inline crates/bevy_ecs/src/event.rs:426]: I do like that this impl is simple and built out of composable pieces, but calling `self.events.get_event(id)` with per-event logic to find the location in the sequence does unnecessary per-event work. The old approach of slicing the "a and b" event queues and chaining them is more efficient. Imo we should port that approach to the ManualEventIteratorWithId impl.

Comment (Aceeri) [reply crates/bevy_ecs/src/event.rs]: This should be the case now

**Review by JoJoJet — APPROVED** [review]
Everything here looks good to me.

*In the future*, I think we should consider renaming `.iter()`. It always feels wrong and surprising that `events.iter()` mutates `events`, and it's especially surprising that `for _ in &events` does not work. Borrowing from `std::io::Read`, I think a potentially better name for that method would be `.read()`.

**Comment by Aceeri** [comment]
I think it somewhat makes sense in that events is basically a saved iterator, we aren't actually mutating `Events<E>` we are mutating the `ManualEventReader`, but shrug.

**Comment by JoJoJet** [comment]
Mutating the event reader makes perfect sense -- I just don't think the method that does it should be called `iter`

**Review by alice-i-cecile — APPROVED** [review]
Can you add the little utility methods (`get_events` and friends) to the change log too? They're quite useful, but easy to overlook.

bors r+

**Pull request closed without merging**